### PR TITLE
ui: surface logical plan in UI

### PR DESCRIPTION
**Pull request by celiala**
We periodically sample and store a serialized representation of
a logical plan's EXPLAIN tree for each unique fingerprint.

This commit focuses on surfacing this plan tree on the UI Statements Page.
Future work might include further iteration on the data's presentation.

Representation of logical plan surfaced on Statements page:
explain tree

Which has the same structure has its EXPLAIN output in the SQL CLI:
explain-sqlcli

Release note (admin ui change): Surface sample logical plans for each
unique fingerprint.

### CODE REVIEW

File: pkg/ui/src/views/statements/statementDetails.tsx [hunk 1]
```diff
@@ -224,13 +225,20 @@ class StatementDetails extends React.Component<StatementDetailsProps, StatementD
...
       <div className="content l-columns">
         <div className="l-columns__left">
           <section className="section section--heading">
             <SqlBox value={ statement } />
           </section>
+          <section className="section">
+            <h3>Logical Plan</h3>
+            {logicalPlan
+            ? <PlanView plan={logicalPlan} />
+            : <p>No plan captured yet.</p>}
```
Comment (vilterp) [inline pkg/ui/src/views/statements/statementDetails.tsx]: Nit: I'd indent the lines starting with `?` and `:`.

But also I've been using this formatter (https://prettier.io/) for the management console and it's saved me a lot of time (and debates like this). Maybe it's time to apply it to the whole codebase (in a separate PR), then have you and @couchand use that going forward.

File: pkg/ui/src/views/statements/planView.tsx [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    </span>
+  );
+}
+
+interface CollapsedPlanAttr {
+  key: string;
+  value: string[];
+}
+
+function collapseRepeatedAttrs(attrs: Attr$Properties[]): CollapsedPlanAttr[] {
```
Comment (vilterp) [inline pkg/ui/src/views/statements/planView.tsx]: would be good to test this.

**Review by vilterp — COMMENTED** [review]
My, what an elegant way of representing plans! (lol jk).

This is sort of a question for Jeremy, but I think the orange (used for keys in the attributes) is pretty hard to read. Another color, preferably from our palette, would be good.

Code looks good, but I haven't been able to see a plan yet locally because it always says "no plan captured yet" — maybe an issue in the backend PR?

**Comment by celiala** [comment]
Appreciate the feedback!

> haven't been able to see a plan yet locally because it always says "no plan captured yet"

Hmm, let me investigate..

> orange (used for keys in the attributes) is pretty hard to read

Good callout! I'll sync with Jeremy on using a color that's easier to read :)



> Code looks good, but I haven't been able to see a plan yet locally because it always says "no plan captured yet" — maybe an issue in the backend PR?

Hmm

**Review by couchand — CHANGES_REQUESTED** [review]
Looks reasonable enough, though I think the generated markup could be made simpler.  Lacking clear style requirements I would err on the side of excessively semantic markup rather than adding generic wrapper elements for class names just to assign colors.

In fact, I'd recommend against color at all, I don't think it's clear that it's necessary, and it makes the display noisy.

> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33483#-:-LVdVagPD6lqxv9QBbRp:b2ozrhr)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/ui/src/views/statements/planView.tsx, line 20 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33483#-LVdQtGV7wZAFr_UxCdZ:-LVdQtGV7wZAFr_UxCd_:b-mfo492) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ace7c5f64fe20ef41a3cd0f01056060d7fb9778/pkg/ui/src/views/statements/planView.tsx#L20)):*
> ```XML
> import { cockroach } from "src/js/protos";
> import { intersperse } from "src/util/intersperse";
> import PlanNode$Properties = cockroach.sql.IExplainTreePlanNode;
> ```

The dollar-sign forms sprinkled about the codebase are an artifact of an old version of ProtobufJS, let's just use the new name instead.
___
*[pkg/ui/src/views/statements/planView.tsx, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33483#-LVdRBAeAT1KxBPPLNIy:-LVdRBAeAT1KxBPPLNIz:b-rz12a2) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ace7c5f64fe20ef41a3cd0f01056060d7fb9778/pkg/ui/src/views/statements/planView.tsx#L25)):*
> ```XML
> export function PlanView(props: { plan: PlanNode$Properties }) {
>   return (
>     <div className="plan-view">
> ```

Looks like there's no need for a wrapper div, just name the content element instead.
___
*[pkg/ui/src/views/statements/planView.tsx, line 27 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33483#-LVdRVnN-j6Z6WGzIuSg:-LVdRVnN-j6Z6WGzIuSh:b-l2w8zp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ace7c5f64fe20ef41a3cd0f01056060d7fb9778/pkg/ui/src/views/statements/planView.tsx#L27)):*
> ```XML
>     <div className="plan-view">
>       <ul>
>         <li>
> ```

I suspect this singleton list is to add a bullet next to the root plan node so it visually "matches" the nested nodes.  It doesn't make semantic sense, I doubt we'll keep the bulleted list look for long, and I don't like how it uses up a level of nesting anyway, so my preference would be to not add the excess list here.
___
*[pkg/ui/src/views/statements/planView.tsx, line 39 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33483#-LVdS3Ho2z-XHu0FGLPq:-LVdS3Ho2z-XHu0FGLPr:bxzylxs) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ace7c5f64fe20ef41a3cd0f01056060d7fb9778/pkg/ui/src/views/statements/planView.tsx#L39)):*
> ```XML
> }
> 
> function PlanNode(props: PlanNodeProps): React.ReactElement<PlanNodeProps> {
> ```

I wouldn't expect to need to annotate the return type.
___
*[pkg/ui/src/views/statements/planView.tsx, line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33483#-LVdSG441w2rHJmfPQbS:-LVdSG441w2rHJmfPQbT:b7umbex) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ace7c5f64fe20ef41a3cd0f01056060d7fb9778/pkg/ui/src/views/statements/planView.tsx#L41)):*
> ```XML
> function PlanNode(props: PlanNodeProps): React.ReactElement<PlanNodeProps> {
>   const node = props.node;
>   node.children = node.children || [];
> ```

Mutation is evil!  Here it kind of makes sense (we're regularizing the external input because the ProtobufJS generated API is annoying), but I'd much rather do that when we get the data rather than sneaking it inside a render.
___
*[pkg/ui/src/views/statements/planView.tsx, line 47 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33483#-LVdSggw4VKRjGUY1lGQ:-LVdSggw4VKRjGUY1lGR:b-6b445i) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ace7c5f64fe20ef41a3cd0f01056060d7fb9778/pkg/ui/src/views/statements/planView.tsx#L47)):*
> ```XML
>     <div className="plan-node">
>       <span className="plan-node__name">{node.name}</span>
>       <span className="plan-node__attrs">
> ```

We could use semantic markup here, I think a `dl` would be appropriate.
___
*[pkg/ui/src/views/statements/planView.tsx, line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33483#-LVdTPoBB24pd27FHbiu:-LVdTPoBB24pd27FHbiv:bf85hh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ace7c5f64fe20ef41a3cd0f01056060d7fb9778/pkg/ui/src/views/statements/planView.tsx#L49)):*
> ```XML
>       <span className="plan-node__attrs">
>         {collapsedAttrs.map((attr) => (
>           <span className="plan-node__attr" key={attr.key}>
> ```

BEM is theoretically the system we use, which I think would suggest that each attr be a block and the key and value be elements.
___
*[pkg/ui/src/views/statements/planView.tsx, line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33483#-LVdThYCDYNibQyG_Vcu:-LVdThYCDYNibQyG_Vcv:b-ox7bxj) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ace7c5f64fe20ef41a3cd0f01056060d7fb9778/pkg/ui/src/views/statements/planView.tsx#L52)):*
> ```XML
>             <span className="plan-node__attr-key">{attr.key}</span>
>             <span className="plan-node__attr-eq">=</span>
>             {typeof attr.value === "string"
> ```

This type switch makes me sad, but I suppose that ship has already sailed?  In any case, can we replace the inline ternary with a self-documenting helper function?
___
*[pkg/ui/src/views/statements/planView.tsx, line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33483#-LVdTd33C5iq28Oe-swH:-LVdTd33C5iq28Oe-swI:b-t9iq3z) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ace7c5f64fe20ef41a3cd0f01056060d7fb9778/pkg/ui/src/views/statements/planView.tsx#L54)):*
> ```XML
>             {typeof attr.value === "string"
>               ? <span className="plan-node__attr-value">{attr.value}</span>
>               : renderAttrValueList(attr.value as string[])}
> ```

This type assertion makes me nervous, what else could it be?
___
*[pkg/ui/src/views/statements/planView.tsx, line 58 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33483#-LVdU1jKDjHDGGKTx-sW:-LVdU1jKDjHDGGKTx-sX:b-8o8d76) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ace7c5f64fe20ef41a3cd0f01056060d7fb9778/pkg/ui/src/views/statements/planView.tsx#L58)):*
> ```XML
>         ))}
>       </span>
>       <ul>
> ```

I think this is semantically an ordered list.
___
*[pkg/ui/src/views/statements/planView.tsx, line 71 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33483#-LVdV6br4VD403I7DTAo:-LVdV6br4VD403I7DTAp:b-kd4tx1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ace7c5f64fe20ef41a3cd0f01056060d7fb9778/pkg/ui/src/views/statements/planView.tsx#L71)):*
> ```XML
> function renderAttrValueList(values: string[]) {
>   return (
>     <span className="plan-node__attr-value-list">
> ```

This method throws in a bunch of extra `span`s and nesting for what seems to be purely textual content.
___
*[pkg/ui/src/views/statements/planView.tsx, line 89 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33483#-LVKdvKH8euVq47Ryjds-r1-89:-LVdUsZ3BVfy2gbhrq-2:bfx3g9t) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ace7c5f64fe20ef41a3cd0f01056060d7fb9778/pkg/ui/src/views/statements/planView.tsx#L89)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

would be good to test this.
</blockquote></details>

Or maybe just use `_.groupBy` instead?
___
*[pkg/ui/src/views/statements/statementDetails.tsx, line 237 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33483#-LVd_F-lEPduVd753rDv:-LVd_F-lEPduVd753rDw:bqm4vjq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ace7c5f64fe20ef41a3cd0f01056060d7fb9778/pkg/ui/src/views/statements/statementDetails.tsx#L237)):*
> ```XML
>           </section>
>           <section className="section">
>             <h3>Logical Plan</h3>
> ```

Since this markup was written the headings have been removed from the other sections in favor of titling within the tables themselves.  I think it would be nice if this heading looked consistent.
___
*[pkg/ui/src/views/statements/statementDetails.tsx, line 240 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33483#-LVKdvKH8euVq47Ryjdt-r1-240:-LVdONyU5skYZSl0OqxN:br1gya8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ace7c5f64fe20ef41a3cd0f01056060d7fb9778/pkg/ui/src/views/statements/statementDetails.tsx#L240)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

Nit: I'd indent the lines starting with `?` and `:`.

But also I've been using this formatter (https://prettier.io/) for the management console and it's saved me a lot of time (and debates like this). Maybe it's time to apply it to the whole codebase (in a separate PR), then have you and @couchand use that going forward. 
</blockquote></details>

I think ternaries are one of the worst inventions of humanity, and I fault JSX for bringing them back from the edge of extinction.

Suggest just adding to the top of `PlanView`:

```
if (!props.plan) {
    return <p>No plan captured yet.</p>
}
```

**Review by celiala — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33483)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/ui/src/views/statements/planView.tsx, line 20 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33483#-LVdQtGV7wZAFr_UxCdZ:-LVdetjg4fzJ26iAOYVz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ace7c5f64fe20ef41a3cd0f01056060d7fb9778/pkg/ui/src/views/statements/planView.tsx#L20)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

The dollar-sign forms sprinkled about the codebase are an artifact of an old version of ProtobufJS, let's just use the new name instead.
</blockquote></details>

Done.
___
*[pkg/ui/src/views/statements/planView.tsx, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33483#-LVdRBAeAT1KxBPPLNIy:-LVdfFpr8oGHNxky8f_p:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ace7c5f64fe20ef41a3cd0f01056060d7fb9778/pkg/ui/src/views/statements/planView.tsx#L25)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Looks like there's no need for a wrapper div, just name the content element instead.
</blockquote></details>

Done.
___
*[pkg/ui/src/views/statements/planView.tsx, line 27 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33483#-LVdRVnN-j6Z6WGzIuSg:-LVdfGEm6yOcQqnKiivW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ace7c5f64fe20ef41a3cd0f01056060d7fb9778/pkg/ui/src/views/statements/planView.tsx#L27)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

I suspect this singleton list is to add a bullet next to the root plan node so it visually "matches" the nested nodes.  It doesn't make semantic sense, I doubt we'll keep the bulleted list look for long, and I don't like how it uses up a level of nesting anyway, so my preference would be to not add the excess list here.
</blockquote></details>

Done.
___
*[pkg/ui/src/views/statements/planView.tsx, line 39 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33483#-LVdS3Ho2z-XHu0FGLPq:-LVdg0MP7z-Uy9bkHDRe:b-qplya4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ace7c5f64fe20ef41a3cd0f01056060d7fb9778/pkg/ui/src/views/statements/planView.tsx#L39)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

I wouldn't expect to need to annotate the return type.
</blockquote></details>

Because I'm referencing <PlanNode /> in the return expression, TypeScript is requesting that I specify the return type here (otherwise it's an implicit `any` type).
___
*[pkg/ui/src/views/statements/planView.tsx, line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33483#-LVdSG441w2rHJmfPQbS:-LVdjnEBBNR-ikhfvkkj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ace7c5f64fe20ef41a3cd0f01056060d7fb9778/pkg/ui/src/views/statements/planView.tsx#L41)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Mutation is evil!  Here it kind of makes sense (we're regularizing the external input because the ProtobufJS generated API is annoying), but I'd much rather do that when we get the data rather than sneaking it inside a render.
</blockquote></details>

Done.
___
*[pkg/ui/src/views/statements/planView.tsx, line 47 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33483#-LVdSggw4VKRjGUY1lGQ:-LVeU6VaCCSh2XFMUwI_:bxr18w0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ace7c5f64fe20ef41a3cd0f01056060d7fb9778/pkg/ui/src/views/statements/planView.tsx#L47)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

We could use semantic markup here, I think a `dl` would be appropriate.
</blockquote></details>

Holding off: we'll get rid of this with the new designs and getting these elements to lay in a similar way via CSS was proving to be a rabbit hole..
___
*[pkg/ui/src/views/statements/planView.tsx, line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33483#-LVdTPoBB24pd27FHbiu:-LVeUh5yBK3BYtlBqaFb:b-53dj31) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ace7c5f64fe20ef41a3cd0f01056060d7fb9778/pkg/ui/src/views/statements/planView.tsx#L49)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

BEM is theoretically the system we use, which I think would suggest that each attr be a block and the key and value be elements.
</blockquote></details>

(same as above)
___
*[pkg/ui/src/views/statements/planView.tsx, line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33483#-LVdThYCDYNibQyG_Vcu:-LVdln6Y-qNG74b7W2c4:b-kwlnca) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ace7c5f64fe20ef41a3cd0f01056060d7fb9778/pkg/ui/src/views/statements/planView.tsx#L52)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

This type switch makes me sad, but I suppose that ship has already sailed?  In any case, can we replace the inline ternary with a self-documenting helper function?
</blockquote></details>

(Removed, see below)
___
*[pkg/ui/src/views/statements/planView.tsx, line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33483#-LVdTd33C5iq28Oe-swH:-LVdlWtv3hGLPFG2zzDI:b3rkzd) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ace7c5f64fe20ef41a3cd0f01056060d7fb9778/pkg/ui/src/views/statements/planView.tsx#L54)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

This type assertion makes me nervous, what else could it be?
</blockquote></details>

Oops, that was leftover from an older commit which no longer applies. Removing.
___
*[pkg/ui/src/views/statements/planView.tsx, line 58 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33483#-LVdU1jKDjHDGGKTx-sW:-LVdlv2r2o0wnJLDUf-d:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ace7c5f64fe20ef41a3cd0f01056060d7fb9778/pkg/ui/src/views/statements/planView.tsx#L58)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

I think this is semantically an ordered list.
</blockquote></details>

Done.
___
*[pkg/ui/src/views/statements/planView.tsx, line 89 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33483#-LVKdvKH8euVq47Ryjds-r1-89:-LVdziXi0v75GsBUzAI8:be4a6qd) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ace7c5f64fe20ef41a3cd0f01056060d7fb9778/pkg/ui/src/views/statements/planView.tsx#L89)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Or maybe just use `_.groupBy` instead?
</blockquote></details>

Done - tests added
___
*[pkg/ui/src/views/statements/statementDetails.tsx, line 237 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33483#-LVd_F-lEPduVd753rDv:-LVdzncz6f7RV-AdgCap:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ace7c5f64fe20ef41a3cd0f01056060d7fb9778/pkg/ui/src/views/statements/statementDetails.tsx#L237)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Since this markup was written the headings have been removed from the other sections in favor of titling within the tables themselves.  I think it would be nice if this heading looked consistent.
</blockquote></details>

Done.
___
*[pkg/ui/src/views/statements/statementDetails.tsx, line 240 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33483#-LVKdvKH8euVq47Ryjdt-r1-240:-LVdnz-B--0IvOAhz3zn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ace7c5f64fe20ef41a3cd0f01056060d7fb9778/pkg/ui/src/views/statements/statementDetails.tsx#L240)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

I think ternaries are one of the worst inventions of humanity, and I fault JSX for bringing them back from the edge of extinction.

Suggest just adding to the top of `PlanView`:

```
if (!props.plan) {
    return <p>No plan captured yet.</p>
}
```
</blockquote></details>

Done.

**Review by celiala — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33483)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/ui/src/views/statements/planView.tsx, line 69 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/33483#-LVeVEiZBeWC5HnVprBG:-LVeVEiZBeWC5HnVprBH:b-q60vuy) ([raw file](https://github.com/cockroachdb/cockroach/blob/42ba189b59943b05249e1cea5d662ccbe3fd05ba/pkg/ui/src/views/statements/planView.tsx#L69)):*
> Because this is referenced by PlanNode's return expression, TypeScript is requesting that I specify the return type here (otherwise it's an implicit any type).

**Comment by celiala** [comment]
Hi there - PTAL, thanks!

Syncing offline with @couchand regarding the semantic feedback: we agreed to not worry about these, as we'll be revisiting these anyway with new designs, and we agreed not to make more work here than needs to be.

**Review by vilterp — APPROVED** [review]

**Comment by celiala** [comment]
Noting that `[TestRemoteDebugModeSetting] status_test.go:890` was causing the build to fail, but this passed locally.

**Pull request merged**